indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/asn1/ObjectIdentifier.kt

```diff
@@ -19,48 +24,125 @@ import kotlinx.serialization.encoding.Encoder
+            if(first().toUByte()>127u) throw Asn1Exception("OID top-level arc can only be number 0, 1 or 2")
+        }
+        nodes?.apply {
+            if (size < 2) throw Asn1StructuralException("at least two nodes required!")
+            if (first() > 2u) throw Asn1Exception("OID top-level arc can only be number 0, 1 or 2")
+            if(first()<2u) {
+                if (get(1) > 39u) throw Asn1Exception("Second segment must be <40")
+            }else {
+                if (get(1) > 47u) throw Asn1Exception("Second segment must be <48")
+            }
+            forEach { if (it.isNegative) throw Asn1Exception("Negative Number encountered: $it") }
+        }
+    }
+
+
+    /**
+     * Efficient, but cursed encoding of OID nodes, see [Microsoft's KB entry on OIDs](https://learn.microsoft.com/en-us/windows/win32/seccertenroll/about-object-identifier)
+     * for details.
+     * Lazily evaluated.
+     */
+    val bytes: ByteArray by if (bytes != null) lazyOf(bytes) else lazy {
+        this.nodes.toOidBytes()
+    }
 
-        if (nodes.first() > 2u) throw Asn1Exception("OID must start with either 1 or 2")
-        if (nodes[1] > 39u) throw Asn1Exception("Second segment must be <40")
+    /**
+     * Lazily evaluated list of OID nodes (e.g. `[1, 2, 35, 4654]`)
+     */
+    val nodes by if (nodes != null) lazyOf(nodes) else lazy {
+        val (first, second) =
+            if (this.bytes[0] >= 80) {
+                BigInteger.fromUByte(2u) to BigInteger.fromUInt(this.bytes[0].toUByte() - 80u)
+            } else {
+                BigInteger.fromUInt(this.bytes[0].toUByte() / 40u) to BigInteger.fromUInt(this.bytes[0].toUByte() % 40u)
+            }
+        var index = 1
+        val collected = mutableListOf(first, second)
+        while (index < this.bytes.size) {
+            if (this.bytes[index] >= 0) {
+                collected += BigInteger.fromUInt(this.bytes[index].toUInt())
+                index++
+            } else {
+                val currentNode = mutableListOf<Byte>()
+                while (this.bytes[index] < 0) {
+                    currentNode += this.bytes[index] //+= parsed
+                    index++
+                }
+                currentNode += this.bytes[index]
+                index++
+                collected += currentNode.decodeAsn1VarBigInt().first
+            }
+        }
+        collected
     }
 
     /**
-     * @param oid in human-readable format (e.g. "1.2.96")
+     * Creates an OID in the 2.25 subtree that requires no formal registration.
+     * E.g. the UUID `550e8400-e29b-41d4-a716-446655440000` results in the OID
+     * `2.25.113059749145936325402354257176981405696`
      */
-    constructor(oid: String) : this(*(oid.split(if (oid.contains('.')) '.' else ' ')).map { it.toUInt() }.toUIntArray())
+    @OptIn(ExperimentalUuidApi::class)
+    constructor(uuid: Uuid) : this(
+        bytes = byteArrayOf((2 * 40 + 25).toUByte().toByte(), *uuid.toBigInteger().toAsn1VarInt()),
```

Comment:
i suspect `(2 * 40 + 25)` is not compile time evaluated and would prefer if it were replaced accordingly

(maybe it's premature optimization?)

Reply:
i would hope, it is compile-time evaluated. Anyways, it is comprehensible, because that is the textbook way of encoding those first two nodes. If i just plug in a constant, then I'd need a comment to de-obfuscate the number

Reply:
also: we are already 1000% (not a typo) faster than my initial approach and 500% faster than before this PR and this single expression will not impact that gain

Reply:
(yes a typo, i checked the numbers:  >2000% faster, and >1000% faster, respectively)